feat(trainer): add dataset and model initializer support to container backend

HKanoje opened the pull request:
## What this PR does / why we need it

This PR implements dataset and model initializer support in the container backend, bringing it to feature parity with the Kubernetes backend. This addresses issue #171 by enabling users to automatically download and prepare datasets and models before training starts.

## Solution Overview

This implementation adds full initializer support to the container backend by:

1. **Running initializer containers before training** - Initializers execute sequentially (dataset first, then model) before any training containers start
2. **Using shared volumes** - All containers (initializers and training nodes) share the same workspace directory on the host
3. **Proper lifecycle management** - Initializers must complete successfully before training begins, with automatic cleanup on failures
4. **Comprehensive error handling** - Clear error messages and proper resource cleanup when initialization fails

## Detailed Changes

### 1. New Utility Functions (`kubeflow/trainer/backends/container/utils.py`)

#### `build_initializer_command(initializer, init_type)`
Builds the appropriate container command based on initializer type:
- **HuggingFace**: Uses `kubeflow.storage_initializer.hugging_face` module
- **S3**: Uses `kubeflow.storage_initializer.s3` module  
- **DataCache**: Uses `kubeflow.storage_initializer.datacache` module

#### `build_initializer_env(initializer, init_type)`
Constructs environment variables from initializer configuration:
- Sets `STORAGE_URI` from the initializer config
- Sets `OUTPUT_PATH` to `/workspace/dataset` or `/workspace/model` based on type
- Adds optional fields like `ACCESS_TOKEN`, `ENDPOINT`, `REGION`, etc.
- Handles DataCache-specific variables like `CLUSTER_SIZE`, `METADATA_LOC`

#### `get_initializer_image()`
Returns the initializer container image (`kubeflow/training-operator:latest`).
This can be made configurable via backend config in future iterations.

### 2. Enhanced ContainerBackend (`kubeflow/trainer/backends/container/backend.py`)

#### `_run_initializers(job_name, initializer, workdir, network_id)`
Orchestrates the initialization phase:
- Pulls the initializer image if needed (respects `pull_policy`)
- Runs dataset initializer if configured
- Runs model initializer if configured
- Ensures proper sequencing and error propagation

#### `_run_single_initializer(job_name, initializer_config, init_type, image, workdir, network_id)`
Executes a single initializer container:
- Creates container with proper labels for tracking
- Mounts shared volume to `/workspace`
- Monitors container status with configurable timeout (default 10 minutes)
- Waits for successful completion (exit code 0)
- Captures and reports logs on failure
- Cleans up failed containers automatically

#### Updated `train()` method
- Creates network and working directory first
- Runs initializers before generating training script
- Only proceeds to training if initialization succeeds
- Maintains backward compatibility (initializers are optional)

#### Updated `__get_trainjob_from_containers()`
- Correctly counts only training nodes for `num_nodes` (excludes initializers)
- Includes initializer containers in the steps list
- Properly tracks initializer status

#### Updated `get_job_logs()`
- Supports fetching logs from specific initializer steps
- When requesting `node-0` logs (default), only shows training container logs
- Can explicitly request initializer logs with `step="dataset-initializer"` or `step="model-initializer"`

### 3. Comprehensive Test Coverage (`kubeflow/trainer/backends/container/backend_test.py`)

Added 11 new test cases covering:

#### Initialization Success Scenarios
- **HuggingFace dataset initializer** - Tests storage_uri parsing and access_token handling
- **S3 dataset initializer** - Tests endpoint, region, and credential configuration
- **HuggingFace model initializer** - Tests model downloads with ignore_patterns
- **S3 model initializer** - Tests S3-compatible storage with authentication
- **Both dataset and model** - Tests sequential execution of both initializers
- **DataCache initializer** - Tests distributed cache configuration with metadata_loc and cluster size

#### Log Retrieval
- Tests getting logs from dataset-initializer step
- Tests getting logs from model-initializer step  
- Tests that default node logs exclude initializer logs

#### Error Handling
- **Non-zero exit code** - Verifies proper error reporting when initializer fails
- **Timeout handling** - Ensures timeout errors are caught and reported
- **Resource cleanup** - Confirms containers and networks are cleaned up on failure

## Implementation Details

### Initialization Flow
1. User calls `train()` with `initializer` parameter

2. ContainerBackend creates:
```
   ├── Working directory: ~/.kubeflow/trainer/containers/<job-name>/
   └── Network: <job-name>-net
```

3. If `initializer.dataset` is set:
```
   ├── Create dataset-initializer container
   ├── Mount workdir to /workspace
   ├── Run initialization (downloads to /workspace/dataset)
   └── Wait for completion (exit code 0)
```

4. If `initializer.model` is set:
```
   ├── Create model-initializer container
   ├── Mount workdir to /workspace
   ├── Run initialization (downloads to /workspace/model)
   └── Wait for completion (exit code 0)
```

5. Create training containers:
```
   ├── Mount same workdir to /workspace
   └── Access data at /workspace/dataset and /workspace/model
```

### Volume Layout

**Host:** `~/.kubeflow/trainer/containers/<job-name>/`
```
├── dataset/ (from dataset-initializer, if configured)
│   └── <downloaded dataset files>
├── model/ (from model-initializer, if configured)
│   └── <downloaded model files>
└── outputs/ (accessible to all training nodes)
    └── <training outputs>
```

**Container Mount:** `/workspace/`
```
├── dataset/ (read by training code)
├── model/ (read by training code)
└── outputs/ (written by training code)
```

## Testing Results

All tests pass with no regressions:

- [x] 43/43 container backend tests passed
- [x] 173/173 trainer module tests passed
- [x] make verify (ruff lint + format) passed
- [x] pre-commit hooks passed

## Usage Example
```python
from kubeflow.trainer import TrainerClient
from kubeflow.trainer.types import (
    CustomTrainer,
    Initializer,
    HuggingFaceDatasetInitializer,
    HuggingFaceModelInitializer,
)

# Define initializers
initializer = Initializer(
    dataset=HuggingFaceDatasetInitializer(
        storage_uri="hf://username/my-dataset",
        access_token="hf_xxxxx",
    ),
    model=HuggingFaceModelInitializer(
        storage_uri="hf://username/my-model",
        access_token="hf_xxxxx",
    ),
)

# Define trainer
trainer = CustomTrainer(
    func=train_func,
    num_nodes=2,
)

# Train with initializers
client = TrainerClient()
job_name = client.train(
    trainer=trainer,
    initializer=initializer,  # <-- Now works with container backend!
)

# Data is automatically available at:
# - /workspace/dataset (in training containers)
# - /workspace/model (in training containers)
```

Fiona-Waters commented on kubeflow/trainer/backends/container/utils.py:
```diff
@@ -211,3 +211,102 @@ def aggregate_container_statuses(adapter, containers: list[dict]) -> str:
+    """
+    # Use the training-operator image which contains initializer scripts
+    # This can be made configurable via backend config in the future
+    return "kubeflow/training-operator:latest"
```
Should we make this configurable rather than hardcoding it?

Fiona-Waters commented on kubeflow/trainer/backends/container/utils.py:
```diff
@@ -211,3 +211,102 @@ def aggregate_container_statuses(adapter, containers: list[dict]) -> str:
+        if isinstance(
+            initializer, (types.HuggingFaceDatasetInitializer, types.HuggingFaceModelInitializer)
+        )
+        else "python -m kubeflow.storage_initializer.datacache "
```
Here we are setting datacache as the default/fallback, do we want to do this? In the[kubernetes backend](https://github.com/kubeflow/sdk/blob/main/kubeflow/trainer/backends/kubernetes/utils.py) we offer 2 options and raise a value error if the type is invalid.

HKanoje replied:
You're right - that was inconsistent with the kubernetes backend. Changed to raise `ValueError` with a clear message listing all supported types instead of defaulting to datacache.

Fiona-Waters commented on kubeflow/trainer/backends/container/backend.py:
```diff
@@ -467,6 +473,156 @@ def _get_job_containers(self, name: str) -> list[dict]:
+            import time
+
+            timeout = 600  # 10 minutes timeout for initialization
+            polling_interval = 2
```
Could using wait API be supported?

Fiona-Waters reviewed (commented):
Thanks @HKanoje for this PR! I've left some comments, ptal! Thanks

HKanoje commented on kubeflow/trainer/backends/container/backend.py:
```diff
@@ -467,6 +473,156 @@ def _get_job_containers(self, name: str) -> list[dict]:
+            # Clean up the failed container
+            from contextlib import suppress
+
+            with suppress(Exception):
```
Implemented `wait_for_container()` in the adapter interface and both Docker/Podman adapters. This replaces the polling loop with a single blocking wait call - much more efficient.

HKanoje commented:
@kramaranya Done! All commits are now signed.

HKanoje commented on kubeflow/trainer/backends/container/backend.py:
```diff
@@ -257,6 +257,26 @@ def train(
+            logger.debug(f"Created network: {network_id}")
+
+            # Run initializers if configured
+            if initializer:
```
Done! Added `_cleanup_container_resources()` helper method in commit 0b7a952 to consolidate the duplicated cleanup logic across exception handlers and delete_job().

Fiona-Waters commented on kubeflow/trainer/backends/container/backend.py:
```diff
@@ -468,6 +505,153 @@ def _get_job_containers(self, name: str) -> list[dict]:
+            environment=env,
+            labels=labels,
+            volumes=volumes,
+            working_dir=constants.WORKSPACE_PATH,
```
This should be /app - https://github.com/kubeflow/trainer/blob/master/cmd/initializers/dataset/Dockerfile#L3
```suggestion
            working_dir=/app,
```

Fiona-Waters commented on kubeflow/trainer/backends/container/types.py:
```diff
@@ -65,3 +65,11 @@ class ContainerBackendConfig(BaseModel):
         description="Configuration for training runtime sources",
     )
+    initializer_image: str = Field(
+        default="kubeflow/training-operator:latest",
```
Those images come from here FYI https://github.com/kubeflow/trainer/tree/master/cmd/initializers

andreyvelich commented on kubeflow/trainer/backends/container/backend.py:
```diff
@@ -196,6 +196,35 @@ def _runtime_type(self) -> str:
+        # Stop and remove containers
+        if container_ids:
+            for container_id in container_ids:
+                with suppress(Exception):
```
Why do we use suppress here? We don't do that in other part of SDK.

andreyvelich commented on kubeflow/trainer/backends/container/backend_test.py:
```diff
@@ -861,3 +886,333 @@ def test_create_adapter_error_message_format():
+
+
+# Tests for Initializer Support
+@pytest.mark.parametrize(
```
Can you add these test cases to the `test_train()` API: https://github.com/kubeflow/sdk/blob/6241fae531be0d7dc23a00481bcfe387f486ec58/kubeflow/trainer/backends/container/backend_test.py#L470

andreyvelich commented on kubeflow/trainer/backends/container/backend_test.py:
```diff
@@ -861,3 +886,333 @@ def test_create_adapter_error_message_format():
+        ),
+    ],
+)
+def test_get_logs_with_initializers(container_backend, test_case):
```
Same question, add it to test_get_job_logs()

andreyvelich commented on kubeflow/trainer/backends/container/backend.py:
```diff
@@ -468,6 +506,157 @@ def _get_job_containers(self, name: str) -> list[dict]:
+                workdir=workdir,
+                network_id=network_id,
+            )
+            logger.debug("Model initializer completed")
```
Is there a way to run these two containers in parallel?
How does multi attach volume works in Docker/Podman?

andreyvelich replied:
Any thoughts @kubeflow/kubeflow-sdk-team @Fiona-Waters ?

andreyvelich commented on kubeflow/trainer/backends/container/backend.py:
```diff
@@ -468,6 +506,157 @@ def _get_job_containers(self, name: str) -> list[dict]:
+        # Run dataset initializer if configured
+        if initializer.dataset:
+            # Get and pull dataset initializer image
+            dataset_image = container_utils.get_initializer_image(self.cfg, "dataset")
```
Can you simplify the logic here to be consistent with what we do in Kubernetes backend: https://github.com/kubeflow/sdk/blob/6241fae531be0d7dc23a00481bcfe387f486ec58/kubeflow/trainer/backends/kubernetes/backend.py#L704

Just simple define two utils function in container utils:
```py
container_utils.get_dataset_initializer()
container_utils.get_model_initializer()
```
Which returns internal type that you can use in the `_adapter.create_and_start_container()` API:
```py
@dataclass
class ContainerInitializer:
   image: str
   command: str
   env: dict
```

WDYT @HKanoje @Fiona-Waters ?

HKanoje commented:
Hi @andreyvelich @Fiona-Waters,

I've addressed all the review comments in the latest commit:

- Updated default images to GHCR (ghcr.io/kubeflow/trainer/dataset-initializer:latest and ghcr.io/kubeflow/trainer/model-initializer:latest)
- Replaced suppress with try-except blocks (added # noqa: SIM105 to satisfy linter)
- Refactored initializer utils with ContainerInitializer dataclass
- Added get_dataset_initializer() and get_model_initializer() functions following kubernetes backend pattern
- Simplified env building using get_optional_initializer_envs() helper
- Removed init_type parameter (now determined from initializer type)
- Removed DataCache support (not supported in container backend)
- Merged initializer tests into test_train() and test_get_job_logs()
- Removed duplicate test functions
